admin/class-judoka-import-handler.php

```diff
@@ -0,0 +1,102 @@
+        $this->judoka_model = new Judoka_Model();
+    }
+
+    public function handle_import() {
+        if (!$this->verify_import_request()) {
+            return;
+        }
+
+        try {
+            $import_count = $this->import_csv($_FILES['judoka_import_file']['tmp_name']);
+            wp_send_json_success([
+                'message' => sprintf('%d judokas successfully imported', $import_count),
+                'count' => $import_count
+            ]);
+        } catch (Exception $e) {
+            wp_send_json_error('Import error: ' . $e->getMessage());
+        }
+    }
+
+    private function verify_import_request() {
+        if (!wp_verify_nonce($_POST['judoka_import_nonce'], 'import_judoka_nonce')) {
+            wp_send_json_error('Invalid nonce');
+            return false;
+        }
+
+        if (!current_user_can('manage_options')) {
+            wp_send_json_error('Insufficient permissions');
+            return false;
+        }
+
+        if (!isset($_FILES['judoka_import_file']) || $_FILES['judoka_import_file']['error'] !== UPLOAD_ERR_OK) {
+            wp_send_json_error('No file uploaded or upload error');
+            return false;
+        }
+
+        $file_ext = strtolower(pathinfo($_FILES['judoka_import_file']['name'], PATHINFO_EXTENSION));
+        if ($file_ext !== 'csv') {
+            wp_send_json_error('Invalid file format. Please upload a CSV file.');
+            return false;
+        }
+
+        return true;
+    }
+
+    private function import_csv($file_path) {
+        if (($handle = fopen($file_path, "r")) === FALSE) {
+            throw new Exception('Unable to open file');
+        }
+
+        $headers = fgetcsv($handle, 0, ",");
+        if (!$headers) {
+            fclose($handle);
+            throw new Exception('Invalid CSV format: no headers found');
+        }
+
+        $required_headers = ['full_name', 'birth_date', 'category', 'weight', 'club', 'grade', 'gender'];
+        $headers = array_map('trim', array_map('strtolower', $headers));
+        $missing_headers = array_diff($required_headers, $headers);
+
+        if (!empty($missing_headers)) {
+            fclose($handle);
+            throw new Exception('Missing required columns: ' . implode(', ', $missing_headers));
+        }
+
+        $column_indexes = array_flip($headers);
+        $import_count = 0;
+
+        while (($data = fgetcsv($handle, 0, ",")) !== FALSE) {
+            try {
+                $judoka_data = $this->prepare_judoka_data($data, $column_indexes);
+                if ($this->judoka_model->create_judoka($judoka_data)) {
+                    $import_count++;
+                }
+            } catch (Exception $e) {
+                continue;
+            }
+        }
+
+        fclose($handle);
+        return $import_count;
+    }
+
+    private function prepare_judoka_data($data, $column_indexes) {
+        return [
+            'full_name' => trim($data[$column_indexes['full_name']]),
```

Comment:
Cette transformation n'est pas robuste.

Reply:
Une approche naïve consiste à isoler le formatage du nom complet dans une méthode statique. Cela permet d'appliquer un formatage plus robuste sans impacter les responsabilités de la méthode. Normalement il faut que le formatage soit tel qu'on détecte facilement les doublons et qu'on les évite.
```php
private function prepare_judoka_data($data, $column_indexes) {
    return [
        'full_name' => Judoka::formatFullName(trim($data[$column_indexes['full_name']])),
        'birth_date' => trim($data[$column_indexes['birth_date']]),
        'category' => trim($data[$column_indexes['category']]),
        'weight' => floatval(trim($data[$column_indexes['weight']])),
        'club' => trim($data[$column_indexes['club']]),
        'grade' => trim($data[$column_indexes['grade']]),
        'gender' => trim($data[$column_indexes['gender']])
    ];
}
```
On peut définir la fonction statique ici ⤵️
```php
class Judoka {
    public static function formatFullName($fullName) {
        // Ajoutez ici le formatage spécifique si nécessaire
        return ucwords(strtolower($fullName)); // Exemple : conversion en majuscule initiale pour chaque mot
    }
}
```

# Pourquoi une telle pratique ?
<html>
<body>
<!--StartFragment--><html><head></head><body><p>Pour éviter la création de doublons d'utilisateurs dans une application, en particulier dans des scénarios où un administrateur saisit manuellement les informations, il est crucial de suivre certaines bonnes pratiques. Voici les recommandations clés pour prévenir les doublons et minimiser les erreurs de saisie :</p>
<hr>
<h3>1. <strong>Normalisation des Données Avant Sauvegarde</strong></h3>
<ul>
<li>Appliquez une normalisation systématique des données avant de les enregistrer dans la base.
<ul>
<li><strong>Convertir en minuscules</strong> : Les noms (<code inline="">full_name</code>, <code inline="">email</code>) doivent être transformés en minuscules pour comparaison.</li>
<li><strong>Supprimer les espaces inutiles</strong> : Enlevez les espaces avant/après les noms.</li>
<li><strong>Retirer les accents</strong> : Les caractères comme <code inline="">é</code> doivent être convertis en <code inline="">e</code>.</li>
</ul>
</li>
</ul>
<p><strong>Exemple PHP :</strong></p>
<pre><code class="language-php">function normalizeInput($input) {
    return strtolower(trim(preg_replace('/\s+/', ' ', iconv('UTF-8', 'ASCII//TRANSLIT', $input))));
}

$normalizedName = normalizeInput("  John SMITH ");
</code></pre>
<hr>
<h3>2. <strong>Validation sur le Champ Email</strong></h3>
<ul>
<li>L'email est généralement unique et constitue un identifiant robuste.</li>
<li>Vérifiez systématiquement que l'email est unique avant la création d'un utilisateur.</li>
</ul>
<p><strong>Requête SQL :</strong></p>
<pre><code class="language-sql">SELECT COUNT(*) FROM users WHERE email = :email;
</code></pre>
<p><strong>Bonnes pratiques :</strong></p>
<ul>
<li>Bloquer la création d’un utilisateur si l’email existe déjà.</li>
<li>Utiliser des contraintes uniques au niveau de la base de données :
<pre><code class="language-sql">ALTER TABLE users ADD CONSTRAINT unique_email UNIQUE (email);
</code></pre>
</li>
</ul>
<hr>
<h3>3. <strong>Recherche de Doublons Basée sur les Noms</strong></h3>
<ul>
<li>Avant de créer un utilisateur, effectuez une recherche dans la base de données pour détecter les entrées similaires.</li>
<li>Combinez plusieurs critères pour minimiser les faux positifs :
<ul>
<li>Nom complet (<code inline="">full_name</code>).</li>
<li>Date de naissance ou autres informations personnelles comme le téléphone.</li>
</ul>
</li>
</ul>
<p><strong>Requête SQL :</strong></p>
<pre><code class="language-sql">SELECT * FROM users 
WHERE LOWER(full_name) = LOWER(:full_name) 
  AND birth_date = :birth_date;
</code></pre>
<hr>
<h3>4. <strong>Utilisation de Méthodes Phonétiques</strong></h3>
<ul>
<li>Utilisez des algorithmes comme <strong>Soundex</strong> ou <strong>Metaphone</strong> pour comparer les noms similaires.</li>
<li>Cela aide à détecter les variations comme <code inline="">"John Smith"</code> et <code inline="">"Jon Smyth"</code>.</li>
</ul>
<p><strong>Exemple PHP :</strong></p>
<pre><code class="language-php">if (soundex("John Smith") === soundex("Jon Smyth")) {
    echo "Potentiel doublon détecté.";
}
</code></pre>
<hr>
<h3>5. <strong>Alertes et Suggestions à l'Administrateur</strong></h3>
<ul>
<li>Lorsqu’un utilisateur est sur le point d’être créé, comparez les données saisies avec les utilisateurs existants et affichez une alerte si des doublons potentiels sont détectés.</li>
<li>Proposez une liste d'utilisateurs similaires pour vérification.</li>
</ul>
<p><strong>Interface utilisateur :</strong></p>
<ul>
<li>Avant validation, affichez :
<ul>
<li>Les utilisateurs avec le même email.</li>
<li>Les utilisateurs avec des noms similaires selon Soundex/Levenshtein.</li>
</ul>
</li>
</ul>
<hr>
<h3>6. <strong>Formulaires Intelligents</strong></h3>
<ul>
<li><strong>Auto-complétion</strong> : Proposez des suggestions en temps réel lorsqu'un administrateur commence à saisir un nom.</li>
<li><strong>Correction des erreurs courantes</strong> :
<ul>
<li>Détectez les erreurs typographiques et proposez des corrections (par exemple, <code inline="">"Jonh"</code> → <code inline="">"John"</code>).</li>
</ul>
</li>
<li><strong>Validation renforcée</strong> : Ajoutez des validations frontend et backend pour signaler les incohérences.</li>
</ul>
<p><strong>Exemple d’auto-complétion :</strong></p>
<ul>
<li>Utiliser une recherche fuzzy (par exemple, via <code inline="">tntsearch</code>) pour afficher les correspondances proches.</li>
</ul>
<hr>
<h3>7. <strong>Contraintes Uniques et Indices dans la Base de Données</strong></h3>
<ul>
<li>Définir des <strong>contraintes uniques combinées</strong> sur les colonnes qui permettent d'identifier un utilisateur.
<ul>
<li>Par exemple, <code inline="">full_name</code> + <code inline="">birth_date</code> + <code inline="">email</code>.</li>
</ul>
</li>
</ul>
<p><strong>SQL :</strong></p>
<pre><code class="language-sql">ALTER TABLE users ADD CONSTRAINT unique_user UNIQUE (full_name, birth_date);
</code></pre>
<hr>
<h3>8. <strong>Processus de Vérification Humaine</strong></h3>
<ul>
<li>Si des doublons sont détectés, imposez une vérification manuelle avant la création.</li>
<li>Ajoutez un workflow pour gérer les doublons potentiels :
<ul>
<li>L'administrateur peut confirmer si l'entrée est un nouvel utilisateur ou une mise à jour.</li>
</ul>
</li>
</ul>
<hr>
<h3>9. <strong>Gestion des Erreurs après Création</strong></h3>
<ul>
<li>Créez un système pour fusionner les utilisateurs en cas de doublons détectés après coup.</li>
<li>Fournissez des outils d'administration pour comparer les entrées et les fusionner.</li>
</ul>
<hr>
<h3>10. <strong>Utilisation d'IA ou de Services Externes</strong></h3>
<ul>
<li>Intégrez des services ou bibliothèques basés sur l’IA pour détecter des doublons complexes.</li>
<li>Exemple : Utilisez des services comme <strong>ElasticSearch</strong> ou <strong>Algolia</strong> pour la recherche fuzzy.</li>
</ul>
<hr>
<h3>Synthèse des Bonnes Pratiques</h3>

Étape | Recommandation
-- | --
Normalisation des données | Minuscule, suppression des espaces et accents
Validation unique | Utilisation de l’email comme identifiant unique
Recherche de doublons | Soundex, Levenshtein, ou comparaison stricte (nom + date de naissance)
Alertes administrateur | Suggérer des correspondances similaires avant validation
Contraintes DB | Contraintes uniques combinées sur full_name + birth_date ou email
Fusion des doublons | Fournir un outil pour fusionner ou corriger les doublons
Auto-complétion | Recherche fuzzy et suggestion d’utilisateurs existants


<p>En combinant ces approches, vous réduirez significativement la probabilité de doublons dus à des erreurs humaines ou des variations dans les données.</p></body></html><!--EndFragment-->
</body>
</html>